### PR TITLE
Update rule 'AllCaps'

### PR DESCRIPTION
**Pull request by riturajj**
'AllCaps' rule breaks out while trying to check for 'Text' elements of None type. Adding 2 lines of code to check if the element is None before we access 'Text' element in it.

### CODE REVIEW

File: rules.py [hunk 1]
```diff
@@ -953,25 +953,30 @@ def check(self, element):
         object_id = element.get("objectId")
         if element.tag == "Candidate":
             ballot_name = element.find("BallotName")
-            if ballot_name.find("Text") is not None:
```
Comment (ccongchen) [inline rules.py]: No need to check here as BallotName is required.

File: rules.py [hunk 2]
```diff
@@ -953,25 +953,30 @@ def check(self, element):
...
-                name = ballot_name.find("Text").text
```
Comment (ccongchen) [inline rules.py]: You can combine everything in one line here - name = element.find("BallotName").find("Text").text

File: rules.py [hunk 3]
```diff
@@ -953,25 +953,30 @@ def check(self, element):
...
-                if name is not None and name == name.upper():
```
Comment (ccongchen) [inline rules.py]: IMHO, this shall be a check for none empty string, you can use not name.strip() instead (of name is not None)

File: rules.py [hunk 4]
```diff
@@ -953,25 +953,30 @@ def check(self, element):
...
+            if ballot_name is not None:
+                if ballot_name.find("Text") is not None:
+                    name = ballot_name.find("Text").text
+                    if name is not None and name == name.upper():
+                        raise base.ElectionWarning(
+                            "Line %d. Candidate %s has name in all upper case letters." % (
+                            element.sourceline, object_id))
         elif element.tag == "Contest":
             name_element = element.find("Name")
             if name_element is not None:
```
Comment (ccongchen) [inline rules.py]: Similarly no need to check as "Name" is required for "Contest" entity.

File: rules.py [hunk 5]
```diff
@@ -953,25 +953,30 @@ def check(self, element):
...
                 name = name_element.text
                 if name is not None and name == name.upper():
```
Comment (ccongchen) [inline rules.py]: Similarly check non empty string instead, I'd recommend using 'not name.strip()'

File: rules.py [hunk 6]
```diff
@@ -953,25 +953,30 @@ def check(self, element):
...
-                    raise base.ElectionWarning("Line %d. Contest %s has name in all upper case letters." % (
+                    raise base.ElectionWarning(
+                        "Line %d. Contest %s has name in all upper case letters." % (
                         element.sourceline, object_id))
         else:
             full_name = element.find("FullName")
-            if full_name.find("Text") is not None:
```
Comment (ccongchen) [inline rules.py]: This is where you shall check! :) As FullName is not required for Person.

File: rules.py [hunk 7]
```diff
@@ -953,25 +953,30 @@ def check(self, element):
...
-                name = full_name.find("Text").text
-                if name is not None and name == name.upper():
```
Comment (ccongchen) [inline rules.py]: Again check that name is a non empty string before all caps.

File: rules.py [hunk 8]
```diff
@@ -941,43 +941,41 @@ def check(self, element):
...
-            name_element = element.find("Name")
-            if name_element is not None:
-                name = name_element.text
-                if name is not None and name == name.upper():
+                name = ballot_name.text
+                if name and name.isupper():
                     raise base.ElectionWarning(
-                        "Line %d. Contest %s has name in all upper case letters." % (
-                        element.sourceline, object_id))
+                        "Line %d. Candidate %s has name in all upper case letters." %
```
Comment (ccongchen) [inline rules.py]: Please correct typo in this case and use Contest Instead.

Comment (riturajj) [reply rules.py]: Actually this particular warning message is for the Candidate, so I included 'Candidate' in the message. The one for 'Contest' is in the next conditional block.

File: rules.py [hunk 9]
```diff
@@ -941,43 +941,41 @@ def check(self, element):
...
             full_name = element.find("FullName")
             if full_name is not None:
-                if full_name.find("Text") is not None:
-                    name = full_name.find("Text").text
-                    if name is not None and name == name.upper():
-                        raise base.ElectionWarning(
-                            "Line %d. Person %s has name in all upper case letters." % (
-                            element.sourceline, object_id))
-
+                name = full_name.find("Text").text
```
Comment (ccongchen) [inline rules.py]: FullName may not be formatted correctly to include the <Text> element, in which case the validator will run into an error.

File: rules.py (hunk 8 above)
Comment (ccongchen) [reply rules.py]: Looks like the code sections before and after this change were misaligned, and you are correct. Thanks for the clarification!

File: rules.py (hunk 1 above)
Comment (riturajj) [reply rules.py]: Updated as per requirement. Added few checks for corner cases as well.

File: rules.py (hunk 2 above)
Comment (riturajj) [reply rules.py]: Updated the conditional checks and covered the corner cases.

File: rules.py (hunk 3 above)
Comment (riturajj) [reply rules.py]: Updated the conditional checks and covered the corner cases.

File: rules.py (hunk 4 above)
Comment (riturajj) [reply rules.py]: Updated the conditional checks and covered the corner cases for Name elements in Contests.

File: rules.py (hunk 5 above)
Comment (riturajj) [reply rules.py]: Updated the conditional checks and covered the corner cases.

File: rules.py (hunk 6 above)
Comment (riturajj) [reply rules.py]: Updated the conditional checks and covered the corner cases for FullName elements.

File: rules.py (hunk 7 above)
Comment (riturajj) [reply rules.py]: Updated the conditional check.

File: rules.py (hunk 9 above)
Comment (riturajj) [reply rules.py]: Updated the conditional check for this.

File: rules.py [hunk 10]
```diff
@@ -941,38 +941,47 @@ def check(self, element):
...
         object_id = element.get("objectId")
         if element.tag == "Candidate":
-            ballot_name = element.find("BallotName")
-            if ballot_name.find("Text") is not None:
-                name = ballot_name.find("Text").text
-                if name is not None and name == name.upper():
-                    raise base.ElectionWarning("Line %d. Candidate %s has name in all upper case letters." % (
-                        element.sourceline, object_id))
+            ballot_name_element = element.find("BallotName")
+            if ballot_name_element is not None:
```
Comment (markcwal-google) [inline rules.py]: I'd suggest getting rid of this nesting by doing the opposite checks, and then returning if true.

e.g.

ballot_name_element = element.find("BallotName")
if ballot_name_element is None:
  return
ballot_name = ballot_name_element.find("Text")
if ballot_name is None:
  return

etc.

File: AllCaps_fail.xml [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+<ElectionReport xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
+<Election>
+<CandidateCollection>
+<Candidate objectId='can-br-alckmin'>
+	<BallotName><Text language='en'>GERALDO ALCKMIN</Text></BallotName>
+</Candidate>
+</CandidateCollection>
+<ContestCollection>
+<Contest xsi:type="CandidateContest" objectId="cc-br1-pres-br">
+	<Name>PRESIDENTIAL ELECTION, ROUND 1, 2018</Name>
```
Comment (ccongchen) [inline AllCaps_fail.xml]: Thanks for adding this test file, Raj. Can you please make sure that we have the schema required elements in this sample? Thank you!

Comment (riturajj) [reply AllCaps_fail.xml]: Schema required elements added to the sample file

**Pull request merged**